File: javascript/test/charge-service.test.js
import _ from 'lodash';

import ChargeService from '../src/app/charge-service';
import exampleData from '../src/config/example-data';
import StatementApi from '../src/lib/statement-api';

describe('ChargeService', () => {
  describe('getRecurringCharges', () => {
    let cardService, statementCharges;

    beforeEach(() => {
      cardService = new ChargeService();

      statementCharges = [];
      jest.spyOn(StatementApi, 'getCharges').mockReturnValue(statementCharges);
    });

    const createCharge = (charge) => {
      statementCharges.push(charge);
    };

    const useExampleData = () => {
      StatementApi.getCharges.mockReturnValue(exampleData.charges);
    };

    test('sums the charges that occur more than once', () => {
      useExampleData();
      const recurringCharges = cardService.getRecurringCharges();

      expect(_.get(recurringCharges, 'Spotify')).toBe(4);
      expect(_.get(recurringCharges, 'Netflix')).toBe(4);
      expect(_.get(recurringCharges, 'HBO Max')).toBe(2);
      expect(_.get(recurringCharges, 'The New York Times Company')).toBe(2);
    });

    test('does not include charges that only occur once', () => {
      useExampleData();
      const recurringCharges = cardService.getRecurringCharges();

      expect(_.get(recurringCharges, 'Lou Malnati\'s')).toBeUndefined();
    });

    describe('when there are multiple one-time charges', () => {
      beforeEach(() => {
        createCharge({
          name: 'Chipotle',
          date: '2021/01/01',
          amount: 9.99,
          user: 'Tom Smith'
        });
        createCharge({
          name: 'Netflix',
          date: '2021/01/01',
          amount: 9.99,
          user: 'Tom Smith'
        });
        createCharge({
          name: 'Netflix',
          date: '2021/02/01',
          amount: 9.99,
          user: 'Tom Smith'
        });
        createCharge({
          name: 'Apple',
          date: '2021/02/01',
          amount: 999.99,
          user: 'John Smith'
        });
      });

      test('does not include charges that only occur once', () => {
        const recurringCharges = cardService.getRecurringCharges();

        expect(_.get(recurringCharges, 'Chipotle')).toBeUndefined();
        expect(_.get(recurringCharges, 'Apple')).toBeUndefined();
      });
    });
  });
});

File: java/src/test/java/com/nextcapital/recruiting/ChargeServiceTest.java
package com.nextcapital.recruiting;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class ChargeServiceTest {

	private List<Map<String, String>> statementCharges;
	private ChargeService service;

	@BeforeEach
	public void beforeFunction() {
		statementCharges = new ArrayList<>();
		StatementApi statementApi = mock(StatementApi.class);
		service = new ChargeService(statementApi);

		when(statementApi.getCharges()).thenReturn(statementCharges);
	}

	@Test
	public void testDoesNotIncludeChargesThatOnlyOccurOnce() {
		addCharge("Netflix", "1/1/2019", "9.99");

		assertFalse(service.getRecurringCharges().contains("Netflix"));
	}

	@Test
	public void testSumsTheChargesThatOccurMoreThanOnce() {
		addCharge("Spotify", "1/1/2019", "9.99");
		addCharge("Spotify", "2/1/2019", "9.99");

		assertTrue(service.getRecurringCharges().get(0).getOccurrance() == 2);
	}

	private void addCharge(String name, String date, String amount) {
		Map<String, String> statementCharge = new HashMap<>();
		statementCharge.put("name", name);
		statementCharge.put("date", date);
		statementCharge.put("amount", amount);
		statementCharges.add(statementCharge);
	}
}
